File: Binary/boatrescue/boatrescue.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/sha.h>

// Secret constant number
#define SECRET_NUMBER 33700



void sha256_hash(const char *str, char outputBuffer[65]) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, str, strlen(str));
    SHA256_Final(hash, &sha256);

    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        sprintf(outputBuffer + (i * 2), "%02x", hash[i]);
    }
    outputBuffer[64] = 0;
}

int main() {
    char secret_hash[65];
    char input_hash[65];
    char input_str[10];

    // Compute the hash of the secret number
    snprintf(input_str, sizeof(input_str), "%d", SECRET_NUMBER);
    sha256_hash(input_str, secret_hash);

    printf("Welcome to the Guess the Number Challenge!\n");
    printf("Guess the secret number!\n");

    while (1) {
        int user_input;
        printf("Enter your guess: ");
        if (scanf("%d", &user_input) != 1 || user_input < 1 || user_input > 40000) {
            printf("Please enter a valid number between 1 and 40000.\n");
            // Clear the input buffer
            while (getchar() != '\n');
            continue;
        }

        // Hash the user's input
        snprintf(input_str, sizeof(input_str), "%d", user_input);
        sha256_hash(input_str, input_hash);

        // Compare hashes
        if (strcmp(input_hash, secret_hash) == 0) {
            char part1[] = "i-CES{9UE5S3d";
            char part2[] = "_C0Ns74Nt_H4S";
            char part3[] = "H3d_NUm63R}";
        
            
            char flag[50];
            snprintf(flag, sizeof(flag), "%s%s%s", part1,part2,part3);
            printf("Congratulations! You guessed it right. The secret number was %d.\n", SECRET_NUMBER);
            printf("The flag is: %s\n", flag);
            break;
        } else {
            printf("https://www.youtube.com/watch?v=cwd_oXaOuNA\n");
            exit(0);
        }
    }

    return 0;
}

File: Binary/boatrescue/boatrescue.py
import hashlib
import sys

# Secret constant number
SECRET_NUMBER = 33700
SECRET_HASH = hashlib.sha256(str(SECRET_NUMBER).encode()).hexdigest()

def hash_input(input_number):
    """Hashes the input number using SHA-256 and returns the hexadecimal digest."""
    return hashlib.sha256(str(input_number).encode()).hexdigest()

def check_guess(user_guess):
    """Checks if the hashed guess matches the secret hash."""
    guessed_hash = hash_input(user_guess)
    return guessed_hash == SECRET_HASH

def main():
    print("Welcome to the Guess the Number Challenge!")
    print("Guess the secret number!")

    # print(f"Hint: The hash of the secret number is: {SECRET_HASH}")

    while True:
        try:
            user_input = int(input("Enter your guess: ").strip())
            if 1 <= user_input <= 40000:
                if check_guess(user_input):
                    print(f"Congratulations! You guessed it right. The secret number was {SECRET_NUMBER}.")
                    print("The flag is: i-CES{9UE5S3d_C0Ns74Nt_H4SH3d_NUm63R}")
                    break
                else:
                    print("https://www.youtube.com/watch?v=1k6y1JvaGyE")
                    sys.exit()
            else:
                print("Please enter a number between 1 and 100.")
        except ValueError:
            print("Invalid input. Please enter a valid number.")

if __name__ == "__main__":
    main()
